_posts/2025-09-23-quarkus-a2a-java-grpc.adoc

```diff
@@ -0,0 +1,346 @@
+List<BiConsumer<ClientEvent, AgentCard>> consumers = getConsumers(messageResponse);
+
+// Create error handler for streaming errors
+Consumer<Throwable> streamingErrorHandler = (error) -> {
+    System.out.println("Streaming error occurred: " + error.getMessage());
+    error.printStackTrace();
+    messageResponse.completeExceptionally(error);
+};
+
+// Create channel factory for gRPC transport
+Function<String, Channel> channelFactory = agentUrl -> {
+    return ManagedChannelBuilder.forTarget(agentUrl).usePlaintext().build();
+};
+
+ClientConfig clientConfig = new ClientConfig.Builder()
+    .setAcceptedOutputModes(List.of("Text"))
+    .build();
+
+// Create the client with both JSON-RPC and gRPC transport support.
+// The A2A server agent's preferred transport is gRPC, since the client
+// also supports gRPC, this is the transport that will get used
+Client client = Client.builder(publicAgentCard) <1>
+    .addConsumers(consumers) <2>
+    .streamingErrorHandler(streamingErrorHandler) <3>
+    .withTransport(GrpcTransport.class, new GrpcTransportConfig(channelFactory)) <4>
+    .withTransport(JSONRPCTransport.class, new JSONRPCTransportConfig()) <5>
+    .clientConfig(clientConfig) <6>
+    .build();
+
+// Create and send the message
+Message message = A2A.toUserMessage(messageText);
+
+System.out.println("Sending message: " + messageText);
+client.sendMessage(message); <7>
+System.out.println("Message sent successfully. Waiting for response...");
+
+try {
+    // Wait for response with timeout
+    String responseText = messageResponse.get();
+    System.out.println("Final response: " + responseText);
+} catch (Exception e) {
+    System.err.println("Failed to get response: " + e.getMessage());
+    e.printStackTrace();
+}
+...
+----
+<1> We can use `Client.builder(publicAgentCard)` to create our A2A client. We need to pass in the `AgentCard`
+for the A2A server agent this client will be communicating with.
+<2> We need to specify event consumers that will be used to handle the responses that will be received from the
+A2A server agent. This will be explained in more detail in the next section.
+<3> The A2A client created by the `Client.builder` will automatically send streaming messages, as opposed to
+non-streaming messages, if it's supported by both the server and the client. We need to specify a handler that will be used for any errors that might occur during streaming.
+<4> We're indicating that we'd like our client to support the gRPC transport.
+<5> We're indicating that we'd like our client to also support the JSON-RPC transport. When communicating with
+an A2A server agent that doesn't support gRPC, this is the transport that would get used.
+<6> We can optionally specify general client configuration and preferences here.
+<7> Once our `Client` has been created, we can send a message to the A2A server agent. The client will automatically
+use streaming if it's supported by both the server and the client.
```

Comment:
Although you've mentioned it earlier, it might be worth emphasising like you do for 5, that if the server doesn't support streaming it will fall back to non-streaming?